### DISCUSSION TITLE
Uninstall command

### DISCUSSION THREAD

**Discussion by MagaKhachatryan** [dokan-dev/dokan-dotnet#324] [Q&A] [answered]
Dear Dokan support team, 
We need a little help here. 
In order to use Dokan in our app we need to install it during our app installation process. So we keep every necessary file of the Dokan library and  in our installer exe, we copy dokan2.sys file into the driver folder, then run the "docanctl.exe /I d" command. Everything works fine. 

But when uninstalling we run "docanctl.exe /r d" command and even after restart. dokan2.sys is not deleted from drivers.  But when uninstalling Dokan library manually from "Apps and Programs" the driver file is deleted. 

So  the question is : Is there a command that will work after the OS restart, or do you have a special command to delete the driver file  that we can use in our uninstaller code?

Thanks in advance!

**Comment by Liryna** [comment, marked as answer]
`dokanctl.exe /r d` will just unregister the driver. You will need to remove the driver yourself. For example by using `MoveFileExA` `MOVEFILE_DELAY_UNTIL_REBOOT` flag which on reboot will remove the file when it is no longer used.

**Reply by MagaKhachatryan** [reply]
Ok, thanks @Liryna very much!
Here I have another question that is very important for us. I will be very glad if you have time for some advice.

In our Mirror folder, we keep encrypted files and we don't create actual folders in Mirror. We keep the needed info for every folder in a unique file. The names of files and folders are created by the hash function to keep everything secure. 

We handle all actions in Mirror to work with our virtual file system encrypted data properly. 
We have a server where files can be modified and we keep it synced with the Desktop client. 

So imagine that the drive is mounted and we are inside a folder "G:\\A".  When this folder is renamed to "G:\\B" in the server, we just delete The info file of folder "A" and create another one for "B ". We must notify Dokan about it so that File Explorer can be refreshed.  The problem is how to notify. 
The call " Dokan.Notify.Rename" with valid paths doesn't work because actual folders don't exist. We tried to notify "Dokan.Notify.Delete"
with the old path, but it doesn't work too. And the File Explorer remains in an already renamed folder.
In the same way, many actions cant be notified, and without refreshing, we can do some actions in already deleted or moved folders that result in bad cases.
So do you have any function that will help us to refresh the File Explorer?

**Reply by MagaKhachatryan** [reply]
We tried to create a Folder with the actual folder name in the Mirror folder (in this case "A" ) and Rename it, but when this is done manually Explorers are updated, when the same steps are done from code (Directory. Create, Directory. Move) explorers aren't updated though Notify. Rename is called.

**Reply by Liryna** [reply]
I am not sure to follow. The mirror you are talking about is really the mirror sample ? or your implementation based on the mirror ?
`Dokan.Notify.Rename` should work. It does not need to have the previous path existing.
Those Notify functions are exactly built to fill the gap when changes come from the server, You should be able to use them whatever your type of files behind is since Explorer only cares and know about what the user sees.
